src/ham/hamctl/subsys-accounts.cpp

```diff
@@ -1,76 +1,213 @@
+typedef enum { normal, escaped, square, getpars } esc_state_t;
+
+#define ESCAPE_NPAR     8
+struct esc_data_c
 {
-    struct termios tty;
-    tcgetattr(STDIN_FILENO, &tty);
-    if( !enable )
-        tty.c_lflag &= ~ECHO;
-    else
-        tty.c_lflag |= ECHO;
+    esc_state_t  state = normal;
+    unsigned     npar  = 0;
+};
+
+/**
+ * @brief Eliminate escape sequences from the input. This may happen, for
+ *  example, when the user presses the up/down/left/right arrows.
+ *
+ *  Here's a list of the sequences supported by this algorithm.
+ *
+ *    Esc 7                 Save Cursor Position
+ *    Esc 8                 Restore Cursor Position
+ *    Esc [ Pn ; Pn ; .. m  Set attributes
+ *    Esc [ Pn ; Pn H       Cursor Position
+ *    Esc [ Pn ; Pn f       Cursor Position
+ *    Esc [ Pn A            Cursor Up
+ *    Esc [ Pn B            Cursor Down
+ *    Esc [ Pn C            Cursor Forward
+ *    Esc [ Pn D            Cursor Backward
+ *    Esc [ Pn G            Cursor Horizontal Absolute
+ *    Esc [ Pn X            Erase Characters
+ *    Esc [ Ps J            Erase in Display
+ *    Esc [ Ps K            Erase in Line
+ *
+ *  Pn is a string of zero or more decimal digits.
+ *  Ps is a selective parameter.
+ */
+static void escape_sequence(esc_data_c & esc, char c)
+{
+    if (esc.state == normal)
+    {
+        if (c == ESC)
+            esc.state = escaped; /* Starting new escape sequence. */
+        return;
+    }
+
+    if (esc.state == escaped)
+    {
+        esc.state = c == '[' ? square : normal;
+        return;
+    }
+
+    if (esc.state == square)
+    {
+        esc.state = getpars;
+        esc.npar = 0;
+        if (c == '?')
+            return;
+    }
+
+    if (esc.state == getpars)
+    {
+        if (c == ';' && esc.npar < (ESCAPE_NPAR - 1))
+        {
+            esc.npar++;
+            return;
+        }
+        if (c >= '0' && c <= '9')
+            return;
+    }
 
-    (void) tcsetattr(STDIN_FILENO, TCSANOW, &tty);
+    esc.state = normal;
 }
 
-static void read_pw(char * pw)
+/**
+ * @brief Read password from stdin. The password gets obscured with '*'
+ *        characters as it is being typed.
+ *
+ * @param[OUT] pw Where password will be saved.
+ */
+void read_pw(char *pw)
 {
-    char      c;
-    unsigned  i = 0;
-    while (std::cin.get(c))
+    char        c;
+    unsigned    i = 0;
+    esc_data_c  esc_data;
+
+    // Disable STDIN echo to obscure password while it is being typed
+    struct termios oflags, nflags;
+
+    tcgetattr(STDIN_FILENO, &oflags);
+    nflags = oflags;
+
+    nflags.c_lflag &= ~(ICANON | ECHO);
+    nflags.c_cc[VTIME] = 0;
+    nflags.c_cc[VMIN] = 1;
+
+    (void)tcsetattr(STDIN_FILENO, TCSANOW/*TCSAFLUSH*/, &nflags);
+
+    while (std::cin.get(c) && (c != '\n'))
     {
-        if (c == '\n')
+        if (esc_data.state != normal)
         {
-            pw[i] = '\0';
+            escape_sequence(esc_data, c);
+            continue;
+        }
+
+        switch (c)
```

Comment:
Do we really need to print `*`? Standard Unix tools (passwd) don't echo anything to the terminal when the user enters a password.

Reply:
I know that standard UNIX don't echo anything. However, I prefer having the visual feedback that characters have been accepted by the utility. Having the * also helps when one has fat fingers (like me) and types more chars than needed.